src/Integration.Vsix.UnitTests/CFamily/ProcessRunnerTests.cs

```diff
@@ -409,11 +410,94 @@ public void ProcRunner_DoNotLogSensitiveData()
             DummyExeHelper.AssertExpectedLogContents(exeLogFile, allArgs);
         }
 
-        #endregion Tests
+        [TestMethod]
+        public void Execute_CancellationTokenIsAlreadyCancelled_ProcessNotExecuted()
+        {
+            var exeName = WriteBatchFileForTest(TestContext,
+                @"@echo Hello world
+xxx yyy
+@echo Testing 1,2,3...>&2
+");
+
+            var logger = new TestLogger();
+            var args = new ProcessRunnerArguments(exeName, true);
+            var runner = CreateProcessRunner(logger);
+
+            args.CancellationToken = new CancellationToken(true);
+
+            // Act
+            var success = runner.Execute(args);
+
+            // Assert
+            success.Should().BeFalse("");
+            runner.ExitCode.Should().Be(0, "Unexpected exit code");
+
+            logger.AssertNoOutputMessages(); 
+        }
+
+        [TestMethod]
+        public async Task Execute_CancellationTokenCancelledMidway_ProcessKilled()
+        {
+            var exeName = WriteBatchFileForTest(TestContext,
+                @"
+@echo Hello world
+waitfor /t 2 somethingThatNeverHappen
+@echo Done!
+");
+
+            var logger = new TestLogger();
+            var cancellationTokenSource = new CancellationTokenSource();
+            var args = new ProcessRunnerArguments(exeName, true)
+            {
+                TimeoutInMilliseconds = 4000,
+                CancellationToken = cancellationTokenSource.Token
+            };
+
+            var runner = CreateProcessRunner(logger);
+
+            bool? result = null;
+
+            var runProcess = Task.Run(() =>
+            {
+                result = runner.Execute(args);
+            });
+
+            cancellationTokenSource.CancelAfter(500);
+
+            await runProcess;
+
+            result.Should().BeFalse("Expecting the process to have failed");
+            runner.ExitCode.Should().Be(-1, "Unexpected exit code");
+            logger.AssertOutputStringExists("Hello world");
+            logger.AssertPartialOutputStringDoesNotExist("Done!");
+        }
+
+        [TestMethod]
+        public void Execute_CancellationTokenCancelledAfterProcessAlreadyFinished_DoesNotThrow()
```

Comment:
@duncanp-sonar there are actually 2 scenarios here, and I only test the first:
1. Process finished completely, and then Cancel is called --> Cancel callback won't be executed at all, since the Register callback is disposed (which is what happens in this test).
2. Process finished, but Register is not yet disposed.

I don't know how to test for the second one. Got an idea? :/ 
3. Process

Reply:
Not really. You don't have a way of pausing the test code between those two points.
You could partially test some of the code by making the `KillProcess` methods `internal static` and testing it.